File: 2D Dynamic Programming/3068. Find the Maximum Sum of Node Values/3068.cpp
#pragma GCC optimize("O3","Oz","unroll-loops","inline-functions","fast-math","march=native","lto","omit-frame-pointer")
class Solution {
public:
    long long maximumValueSum(vector<int>& nums, int k, vector<vector<int>>& edges) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);

        int n=nums.size();
        vector<vector<long long>>dp(n+1, vector<long long>(2,0));
        dp[n][0]=0;
        dp[n][1]=INT_MIN;

        for(int i=n-1; i>=0; --i){
            for(int j=0; j<2; ++j){
                long long change = dp[i+1][j^1]+(nums[i]^k);
                long long noChange = dp[i+1][j]+(nums[i]);
                dp[i][j] = max(change, noChange);
            }
        }
        return dp[0][0];
    }
};

File: 2D Dynamic Programming/3068. Find the Maximum Sum of Node Values/3068.py
class Solution:
    def maximumValueSum(self, nums: List[int], k: int, edges: List[List[int]]) -> int:
        n=len(nums)
        dp=[[0]*2 for _ in range(n+1)]
        dp[n][1]=-float('inf') # odd number is invalid since edge operations are pairs
        dp[n][0]=0

        for i in range(n-1,-1,-1):
            for j in range(2):
                change = dp[i+1][j^1] + (nums[i]^k)
                noChange = dp[i+1][j] + nums[i]
                dp[i][j] = max(change, noChange)
        
        return dp[0][0] # at the end, we need even number of operations since edges are pairs
        